### PR TITLE
Dynamic NPC spawn anywhere on overmap - not just near player

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
```SUMMARY: Features "Dynamic NPC spawn anywhere on overmap - not just near player"```

#### Purpose of change
has the same changes as #35124 as it relies on that PR
Dynamic NPC spawn anywhere on overmap - not just near player
Also instead of trying once to find the nearest instance of a random choice of "safe" location,
find the nearest occurence of any of them, this is smarter behaviour for the NPC to find a spot to travel to.

#### Describe the solution
Modify game.cpp `perhaps_add_random_npc()` to pick a random spot on the overmap, instead of on the edge of players reality bubble.
Modify npcmove.cpp and overmapbuffer.cpp to work with the list of locations, pass them to overmap::find_closest() to accept a vector, and find the closest occurence of any of them.

#### Describe alternatives you've considered
N/A

#### Testing
Debug output for when a dynamic NPC spawns, see where it paths to,
it now paths to the nearest farm/cave/source of safety, instead of choosing a random one miles away, and giving up if that particular random choice dosnt exist nearby, and reverting to another location for the next need.

#### Additional context
Also fixed a little bug where NPCs can teleport to the first point on their omt_path if they get distracted by a fight, then go back to travelling.

And serialized npc omt_path properly

### CODE REVIEW

File: src/game.cpp [hunk 1]
```diff
@@ -4271,7 +4271,7 @@ void game::monmove()
 void game::overmap_npc_move()
 {
     std::vector<npc *> travelling_npcs;
-    for( auto &elem : overmap_buffer.get_npcs_near_player( 75 ) ) {
+    for( auto &elem : overmap_buffer.get_npcs_near_player( 121 ) ) {
```
Comment (cainiaowu) [inline src/game.cpp]: I think we might need to make these magic number into parameters.

File: src/game.cpp [hunk 2]
```diff
@@ -11113,9 +11117,7 @@ void game::perhaps_add_random_npc()
     }
 
     float density = get_option<float>( "NPC_DENSITY" );
-    // TODO: This is inaccurate when the player is near a overmap border, and it will
-    //immediately spawn new npcs upon entering a new overmap. Rather use number of npcs *nearby*.
-    const int npc_num = get_cur_om().get_npcs().size();
+    const int npc_num = overmap_buffer.get_npcs_near_player( 60 ).size();
```
Comment (cainiaowu) [inline src/game.cpp]: magic number here.

File: src/game.cpp [hunk 3]
```diff
@@ -11124,38 +11126,21 @@ void game::perhaps_add_random_npc()
...
-    if( is_river_or_lake( oter ) ) {
-        return;
+    bool spawn_allowed = false;
+    tripoint spawn_point;
+    int counter = 0;
+    while( !spawn_allowed ) {
+        if( counter >= 10 ) {
+            return;
+        }
+        spawn_point = tripoint( rng( 0, 179 ), rng( 0, 179 ), 0 );
```
Comment (cainiaowu) [inline src/game.cpp]: magic number here.

Comment (davidpwbrown) [reply src/game.cpp]: This is the size of an overmap in OMT co-ords.

**Comment by davidpwbrown** [comment]
Oops I was gonna fix this up on some advice from ralreegorganon in the discord, ill do that in a seperate PR, also it has the same commits as #35124 so I guess I can close that one

**Pull request merged**